File: packages/sdk-py/src/blamr_sdk/trace.py
"""@blamr_trace decorator — auto-emit causal edges with previews and lineage."""

from __future__ import annotations

import functools
import time
from contextvars import ContextVar
from typing import Any, Callable, TypeVar

from blamr_sdk.client import BlamrEmitter, _truncate
from blamr_sdk.lineage import HopLineageRegistry, preview_from_value

F = TypeVar("F", bound=Callable[..., Any])

_trace_stack: ContextVar[list[dict[str, Any]]] = ContextVar("blamr_trace_stack", default=[])


def _parent_agent(emitter: BlamrEmitter, agent: str, from_agent: str | None) -> str:
  stack = _trace_stack.get()
  if from_agent:
    return from_agent
  if stack:
    return stack[-1]["agent"]
  return emitter.agent_id


def blamr_trace(
  emitter: BlamrEmitter,
  *,
  agent: str | None = None,
  from_agent: str | None = None,
  call_type: str = "Tool call",
  model: str = "unknown",
) -> Callable[[F], F]:
  """
  Decorator that wraps a function and emits one CausalEdge per invocation.

  Nested decorated calls build parent→child topology automatically.
  """

  def decorator(fn: F) -> F:
    hop_agent = agent or emitter.agent_id

    @functools.wraps(fn)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
      if not emitter.run_id:
        emitter.start_run()

      stack = _trace_stack.get()
      registry: HopLineageRegistry = stack[-1]["registry"] if stack else HopLineageRegistry()
      source_hop_ids = registry.detect_sources(args, kwargs)
      parent = _parent_agent(emitter, hop_agent, from_agent)

      frame = {"agent": hop_agent, "registry": registry}
      _trace_stack.set(stack + [frame])

      start = time.time()
      input_preview = preview_from_value(args[0] if len(args) == 1 and not kwargs else (args, kwargs))

      try:
        result = fn(*args, **kwargs)
        output_preview = preview_from_value(result)
        edge_id = f"edge_{int(time.time() * 1000)}_{id(result) & 0xFFFF:x}"
        fields: dict[str, Any] = {
          "id": edge_id,
          "from_agent": parent,
          "to_agent": hop_agent,
          "call_type": call_type,
          "model": model,
          "latency_ms": int((time.time() - start) * 1000),
          "confidence_out": 0.85 if output_preview else 0.5,
        }
        if input_preview:
          fields["input_preview"] = _truncate(input_preview)
        if output_preview:
          fields["output_preview"] = _truncate(output_preview)
        if source_hop_ids:
          fields["source_hop_ids"] = source_hop_ids
        emitter.emit_edge(**fields)
        registry.register(result, edge_id)
        return result
      except Exception as exc:
        emitter.emit_edge(
          from_agent=parent,
          to_agent=hop_agent,
          call_type=call_type,
          model=model,
          latency_ms=int((time.time() - start) * 1000),
          confidence_out=0.2,
          intent_delta=-0.5,
          output_preview=_truncate(f"error: {exc}"),
          source_hop_ids=source_hop_ids or None,
        )
        raise
      finally:
        _trace_stack.set(stack)

    return wrapper  # type: ignore[return-value]

  return decorator

File: packages/sdk-ts/src/trace.ts
import type { CallType, CausalEdge } from '@blamr/types';
import { HopLineageRegistry, previewFromValue } from './lineage';
import { computeConfidenceOut } from './signals';

export interface BlamrTraceEmitter {
  getCurrentRunId(): string | null;
  startRun(runId?: string): string;
  emitEdge(edge: Partial<CausalEdge>): Promise<void>;
  getDefaultAgentId(): string;
}

export interface BlamrTraceOptions {
  agent?: string;
  fromAgent?: string;
  callType?: CallType;
  model?: string;
}

interface TraceFrame {
  agent: string;
  registry: HopLineageRegistry;
}

const traceStack: TraceFrame[] = [];

function parentAgent(emitter: BlamrTraceEmitter, options: BlamrTraceOptions): string {
  const parent = traceStack[traceStack.length - 1];
  return options.fromAgent ?? parent?.agent ?? emitter.getDefaultAgentId();
}

async function emitTracedHop(
  emitter: BlamrTraceEmitter,
  options: BlamrTraceOptions,
  start: number,
  sourceHopIds: string[],
  inputPreview: string | undefined,
  result: unknown,
): Promise<void> {
  const agent = options.agent ?? emitter.getDefaultAgentId();
  const fromAgent = parentAgent(emitter, options);
  const outputPreview = previewFromValue(result);
  const text = typeof result === 'string' ? result : outputPreview;
  const edgeId = `edge_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;

  await emitter.emitEdge({
    id: edgeId,
    from_agent: fromAgent,
    to_agent: agent,
    call_type: options.callType ?? 'Tool call',
    model: options.model ?? 'unknown',
    latency_ms: Date.now() - start,
    confidence_out: computeConfidenceOut({
      text: typeof text === 'string' ? text : undefined,
      callType: options.callType ?? 'Tool call',
    }),
    ...(inputPreview ? { input_preview: inputPreview } : {}),
    ...(outputPreview ? { output_preview: outputPreview } : {}),
    ...(sourceHopIds.length ? { source_hop_ids: sourceHopIds } : {}),
  });

  const frame = traceStack[traceStack.length - 1];
  frame?.registry.register(result, edgeId);
}

/** Run a zero-arg function as a traced causal hop. */
export async function runTraced<T>(
  emitter: BlamrTraceEmitter,
  options: BlamrTraceOptions,
  fn: () => T | Promise<T>,
): Promise<T> {
  if (!emitter.getCurrentRunId()) emitter.startRun();

  const agent = options.agent ?? emitter.getDefaultAgentId();
  const registry = traceStack[traceStack.length - 1]?.registry ?? new HopLineageRegistry();
  traceStack.push({ agent, registry });

  const start = Date.now();
  try {
    const result = await fn();
    await emitTracedHop(emitter, options, start, [], undefined, result);
    return result;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    await emitter.emitEdge({
      from_agent: parentAgent(emitter, options),
      to_agent: agent,
      call_type: options.callType ?? 'Tool call',
      model: options.model ?? 'unknown',
      latency_ms: Date.now() - start,
      confidence_out: 0.2,
      intent_delta: -0.5,
      output_preview: `error: ${msg.slice(0, 200)}`,
    });
    throw err;
  } finally {
    traceStack.pop();
  }
}

/**
 * Wrap a function so each invocation emits a causal edge with previews and lineage.
 *
 * @example
 * const research = blamrTrace(emitter, { agent: 'researcher' }, async (q) => ...);
 */
export function blamrTrace<TArgs extends unknown[], TResult>(
  emitter: BlamrTraceEmitter,
  options: BlamrTraceOptions,
  fn: (...args: TArgs) => TResult | Promise<TResult>,
): (...args: TArgs) => Promise<TResult> {
  return async (...args: TArgs): Promise<TResult> => {
    if (!emitter.getCurrentRunId()) emitter.startRun();

    const agent = options.agent ?? emitter.getDefaultAgentId();
    const registry = traceStack[traceStack.length - 1]?.registry ?? new HopLineageRegistry();
    const sourceHopIds = registry.detectSources(args);
    traceStack.push({ agent, registry });

    const start = Date.now();
    const inputPreview = previewFromValue(args.length === 1 ? args[0] : args);

    try {
      const result = await fn(...args);
      await emitTracedHop(emitter, options, start, sourceHopIds, inputPreview, result);
      return result;
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      await emitter.emitEdge({
        from_agent: parentAgent(emitter, options),
        to_agent: agent,
        call_type: options.callType ?? 'Tool call',
        model: options.model ?? 'unknown',
        latency_ms: Date.now() - start,
        confidence_out: 0.2,
        intent_delta: -0.5,
        output_preview: `error: ${msg.slice(0, 200)}`,
        ...(sourceHopIds.length ? { source_hop_ids: sourceHopIds } : {}),
      });
      throw err;
    } finally {
      traceStack.pop();
    }
  };
}
